### DISCUSSION TITLE
Towards a library

### DISCUSSION THREAD

**Discussion by TieSKey** [mxgmn/MarkovJunior#52] [Ideas]
Hey, first of all, thx to all the project contributors/sponsors. It's great and I'm already making good use of it.

Going forward I think it would be interesting to turn the project into a proper library. With that goal in mind let me list some high level tasks I think are necessary 
* Separate [Program.cs](https://github.com/mxgmn/MarkovJunior/blob/main/source/Program.cs) from the "core" code into a different project (same solution).
* Move code/resources only used by [Program.cs](https://github.com/mxgmn/MarkovJunior/blob/main/source/Program.cs) to this new project.
* Replace all Console.Write with error messages by Exceptions with the same messages. 
* Handle previous exceptions instead of using (null, -1,-1,-1) as error.
* Make the resources folder easier to change.
* Separate model "loading/parsing" from model "initialization" (could help with the above too).

Additionally, using a different PriorityQueue (or actually copying the .net6 implementation) could allow the project to run in older .net releases.

I'm not touching documentation and syntax topics since those are already being discussed in other posts.

(I've already made some of this changes for my own use, I can contribute them if u want)

**Comment by mxgmn** [comment]
Hi! The main thing that is needed to turn MJ into a good library is for me to use it as a library in some other project. Then I could better understand what is required from a convenient library. I actually plan to do this, but later, maybe this winter.

**Reply by TieSKey** [reply]
Looking forward to it. 
I'm already using it as a library. The main pain points were the lack of exceptions and the difficulty to set/change the base resources directory path.
